### PR TITLE
perf: Delay SignalRelay connection to when a callback is connected

### LINKED ISSUE #276: [idea] Connect SignalRelay to signals only when have callback
### Description

Currently, SignalRelay is always connected to all signals in SignalGroup. It means that emitting an event always triggers some callback. 

Maybe SignalRelay should connect itself to all signals in a group only if something is connected to it itself?

### PR DESCRIPTION
**Pull request by tlambert03**
closes #276

### CODE REVIEW

File: src/psygnal/_group_descriptor.py [hunk 1]
```diff
@@ -268,7 +268,7 @@ def _setattr_and_emit_(self: object, name: str, value: Any) -> None:
 
             # don't emit if the signal doesn't exist or has no listeners
             signal: SignalInstance = group[name]
-            if len(signal) < 2 and not len(group._psygnal_relay):
+            if len(signal) < 1 and not len(group._psygnal_relay):
```
Comment (Czaki) [inline src/psygnal/_group_descriptor.py:271]:
```suggestion
            if len(signal) < 1:
```

**Review by Czaki — COMMENTED** [review]
Generally looks ok. One suggestion and one question about the possibility of small improvement.

File: src/psygnal/_group_descriptor.py (hunk 1 above)
Comment (tlambert03) [reply src/psygnal/_group_descriptor.py:271]: oh how nice! :)

File: src/psygnal/_group.py [hunk 2]
```diff
@@ -63,10 +67,15 @@ def __init__(
         self._signals = signals
         self._sig_was_blocked: dict[str, bool] = {}
 
+    def _append_slot(self, slot: WeakCallback) -> None:
+        super()._append_slot(slot)
+        self._connect_relay()
```
[the remark this answers is not in the document]
Comment (tlambert03) [reply src/psygnal/_group.py]: added in 4fdec36

**Review by Czaki — APPROVED** [review]

**Pull request merged**